[OptimistLabyrinth] 🚀 2단계 - 연관 관계 매핑

OptimistLabyrinth opened the pull request.

Commit:
refactor: put all exception classes into 'exceptions' directory

Commit:
refactor: 테스트 코드에서 size() 대신에 hasSize() 로 검증하기

Commit:
feat: 네 개의 엔티티 사이에 연관관계 표현하기

`@OneToMany`, `@ManyToOne`, `@JoinColumn` 적절하게 사용하기

Commit:
feat: 엔티티 관계를 형성하는 List<> 컬렉션 초기화

Commit:
test: domain 테스트 코드 오류 수정

테스트 중에 static 객체 대신에 로컬 객체 사용하도록 수정

Commit:
feat: QnaService 통합테스트 코드 오류 수정

Commit:
test: 테스트 실행에 앞서서 setup 하는 코드를 개별 테스트 코드마다 다르게 설정

모든 테스트 케이스에 대해서 공통적인 로직으로 셋업을 실행하면
어떤 테스트 케이스에서 어떤 데이터를 사용하는지
파악하기 어려워지므로
개별 테스트 케이스에서 반드시 필요한 정보만을 다루는 것이 바람직하다

Commit:
test: 테스트 코드는 반드시 public 아니어도 됨

테스트코드에서 불필요하게 사용한 public 키워드를 삭제

OptimistLabyrinth commented:
모든 테스트 케이스에 대해서 공통적인 하나의 `@BeforeEach` 를 적용하는 것이 좋지 않은 이유가
어떤 테스트 케이스에서 어떤 데이터를 다루는지 명확하게 파악하기 어렵기 때문이라고 저는 생각해요.

test fixture 를 사용하면 좋은 점이 여러가지가 있지만
그 중에 하나의 장점은
개별 테스트 케이스마다에서 딱 필요한 데이터만 셋업할 수 있게 되는거라고 저는 생각해요.
test fixture 가 반드시 static 으로 정의되어야 하는지 아닌지 아직 잘 모르겠지만
개인적으로는 뭔가를 구현할 때 static 키워드 최소한으로만 사용하는 스타일을 선호해서
지금처럼 완성했습니다.

mskangg commented on src/main/java/qna/domain/Question.java:
```diff
@@ -61,7 +68,8 @@ public String toString() {
                 ", title='" + title + '\'' +
-                ", writerId=" + writerId +
+                ", writer=" + writer +
+                ", answers=" + answers +
```
질문과 답변을 서로 양방향관계 맺어주셨는데, toString 에서 무한루프에 빠지곘네요 🤔
어떻게 끊어야할까요?

OptimistLabyrinth replied:
수정했습니다.

mskangg commented on src/main/java/qna/domain/User.java:
```diff
@@ -81,6 +89,9 @@ public String toString() {
                 ", name='" + name + '\'' +
                 ", password='" + password + '\'' +
                 ", userId='" + userId + '\'' +
+                ", answers=" + answers +
+                ", questions=" + questions +
+                ", deleteHistories=" + deleteHistories +
```
toString 호출시 모두 다 무한루프 돌것 같습니다,,!

OptimistLabyrinth replied:
수정 완료했습니다

mskangg commented on src/main/java/qna/domain/User.java:
```diff
@@ -26,6 +28,12 @@ public class User extends BaseCreatedAndUpdatedAt {
     private String password;
     @Column(length = 20, nullable = false)
     private String userId;
+    @OneToMany(targetEntity = Answer.class, mappedBy = "writer", fetch = FetchType.LAZY)
+    private final List<Answer> answers = new ArrayList<>();
+    @OneToMany(targetEntity = Question.class, mappedBy = "writer", fetch = FetchType.LAZY)
+    private final List<Question> questions = new ArrayList<>();
+    @OneToMany(targetEntity = DeleteHistory.class, mappedBy = "deletedBy", fetch = FetchType.LAZY)
+    private final List<DeleteHistory> deleteHistories = new ArrayList<>();
```
User에서 모두 다 양방향 관계를 맺어주셨네요!
어떤 이유가 있으셔서 맺어주신걸로 보이는데, 이유 궁금합니다! 🤔

OptimistLabyrinth replied:
생각해보니 관계 설정은 양방향으로 해놓고 실제로 사용할때는 단방향처럼 사용하고 있었네요.
양방향으로 사용하도록 수정했습니다.

OptimistLabyrinth replied:
저는 이 코멘트를 수정해볼만한 부분을 암시적으로 안내하려고 작성하셨다고 생각했지만 혹시 궁금해서 물어보신 경우에 대비해서 답변 드릴께요.

양방향 관게 설정을 하면 논리의 시작이 어디이든 상관없이 JOIN 을 실행할 수 있기 때문에
나중에 기능을 추가하거나 수정하면서 쿼리를 작성하기 편하기 때문에
양뱡향 관계를 설정했어요.

예를 들자면 User - Question 을 서로 양방향 관계 설정을 하면

1) 어떤 사용자가 작성한 모든 질문
2) 어떤 질문에 답변한 모든 사용자

이런식으로 어떤 쿼리이든지 처리할 수 있게 되는 장점이 있습니다.

mskangg commented on src/test/java/qna/service/QnaServiceTest.java:
```diff
@@ -1,97 +1,189 @@
     private AnswerRepository answerRepository;
-
-    @Mock
+    @Autowired
```
mock 에서 SpringBootTest 로 변경했기에 더이상 deleteHistoryService 는 필요해지지 않았네요 
지우셔도 될것 같습니다 :)

OptimistLabyrinth replied:
👍✅

mskangg commented on src/test/java/qna/service/QnaServiceTest.java:
```diff
@@ -1,97 +1,189 @@
+
+        @BeforeEach
+        void setup() {
+            final User user1 = new User("javajigi", "password", "name", "javajigi@slipp.net");
+            final User user2 = new User("sanjigi", "password", "name", "sanjigi@slipp.net");
+            final User user3 = new User("pobi", "password", "name", "pobi@slipp.net");
+            this.user1 = userRepository.save(user1);
+            final List<User> users = new ArrayList<>(2);
+            users.add(user2);
+            users.add(user3);
+            userRepository.saveAll(users);
+
+            final Question question1 = new Question("title1", "contents1").writeBy(this.user1);
+            final Question question2 = new Question("title2", "contents2").writeBy(this.user1);
+            this.question1 = questionRepository.save(question1);
+            questionRepository.save(question2);
+
+            final Answer answer1 = new Answer(this.user1, question1, "Answers Contents1");
+            final Answer answer2 = new Answer(this.user1, question2, "Answers Contents2");
+            final Answer answer3 = new Answer(this.user1, question1, "Answers Contents1-2");
+            question1.addAnswer(answer1);
+            question2.addAnswer(answer2);
+            question1.addAnswer(answer3);
+            final List<Answer> answers = new ArrayList<>(3);
+            answers.add(answer1);
+            answers.add(answer2);
+            answers.add(answer3);
+            final List<Question> questions = new ArrayList<>(3);
+            questions.add(question1);
+            questions.add(question2);
+            answerRepository.saveAll(answers);
+            questionRepository.saveAll(questions);
```
given 절이 매우 기네요!
각 구문별로 하고자하는 행위가 있으니 메소드로 구분지을 수 있어보입니다!
메서드화하여 네이밍을 하고 given 절의 가독성을 높이는 것은 어떠신가요?? 🤔
분리하다보면 재사용할 수도 있어보입니다.

OptimistLabyrinth replied:
✅ 수정했습니다

mskangg reviewed (changes requested):
> 모든 테스트 케이스에 대해서 공통적인 하나의 @BeforeEach 를 적용하는 것이 좋지 않은 이유가
어떤 테스트 케이스에서 어떤 데이터를 다루는지 명확하게 파악하기 어렵기 때문이라고 저는 생각해요.

공감합니다!

> test fixture 를 사용하면 좋은 점이 여러가지가 있지만 그 중에 하나의 장점은 개별 테스트 케이스마다에서 딱 필요한 데이터만 셋업할 수 있게 되는거라고 저는 생각해요.
test fixture 가 반드시 static 으로 정의되어야 하는지 아닌지 아직 잘 모르겠지만 개인적으로는 뭔가를 구현할 때 static 키워드 최소한으로만 사용하는 스타일을 선호해서 지금처럼 완성했습니다.

말씀하신 부분 외에도 test fixture 는 재사용성과 가독성을 좋게할 수 있다고 생각합니다!
혹시 테스트코드에서도 `static 키워드 최소한으로만 사용하는 스타일을 선호`하시는 이유가 있으실까요??
런타임시에는 테스트했던 모든 인스턴스들이 사라지며 영향을 주지 않기에 static 메서드의 여부는 크게 작용하지 않을 것이라 생각됩니다. 🤔

기성님 PR 요청 기다렸습니다! 😁
연관관계 매핑 미션 잘수행해주셨어요~!
몇가지 피드백 남겨드렸으니 반영 부탁드려요!

Commit:
chore: 필요없는 멤버변수(deleteHistoryService)  제거

관련 링크
[https://github.com/next-step/jwp-qna/pull/686/files#r1027077754](https://github.com/next-step/jwp-qna/pull/686/files#r1027077754)

Commit:
fix: toString() 메서드에서 참조 무한루프 제거

toString() 메서드 안에서 참조 무한루프가 생겨날만한 곳마다
id 만 출력하도록 수정

Commit:
feat: 엔티티 사이에 양방향 관계 설정을 제대로 활용하기

엔티티 간에 관계 설정을 양방향으로 해두고서
단방향 설정한 것처럼 사용하고 있었다

양방향 관계 설정을 제대로 활용하도록 수정한다

Commit:
test: setup 메서드 공통 로직 함수로 추출

createUser
createQuestion
createAnswer

Commit:
test: 사용하지 않는 멤버변수 삭제

OptimistLabyrinth reviewed (commented):
피드백 주신 부분 위주로 수정해봤습니다.

위에서 static test fixture 에 대해서 긍정적으로 계속 말씀하시는걸 보니
저도 static test fixture 를 만들고 싶다는 생각은 드는데
평소에 테스트 코드 작성 시에 정적 클래스를 잘 만들지 않다 보니
어떤식으로 만들어야 할지 잘 모르겠습니다.

혹시 예시 하나 보여주실 수 있나요??

mskangg reviewed (approved):
피드백 반영 잘하셨습니다!!

> 위에서 static test fixture 에 대해서 긍정적으로 계속 말씀하시는걸 보니

제가 static을 선호해서 말씀드린것이 아니라,

> 개인적으로는 뭔가를 구현할 때 static 키워드 최소한으로만 사용하는 스타일을 선호해서

기성님께서 non-static을 선호하신다하여 어떤 이유로 선호하시는지 궁금하여 여쭤봤던것이에요!

중요한건, test fixture인 것이지 그게 static인지 non-static인지에 대한 부분은 중요하지 않다고 생각해서요!
정리해서 말씀드리자면, **어떤 이유로 non-static을 선호하시는지 궁금했던 것입니다** 😁

> 평소에 테스트 코드 작성 시에 정적 클래스를 잘 만들지 않다 보니
어떤식으로 만들어야 할지 잘 모르겠습니다.
혹시 예시 하나 보여주실 수 있나요??

지금처럼 private 메서드로 중복되는 given절을 추출해서 사용해도 무방합니다.
요구사항이 추가됨으로 인해 추출한 private 메서드가 다른곳에서도 활용을 해야한다면 그때는 팩토리클래스를 만들어 테스트픽스쳐로 활용하면 될것 같습니다.

현재 `createAnswer()` 메서드를 private으로 선언하였는데, 추후 요구사항이 추가되어 다른 곳에서도 해당 메서드가 필요하다면 `TestAnswerFactory.create()` 와 같이 활용하면 좋겠다 라는 피드백이었어요!

2단계 정말정말 고생하셨습니다 👍 
다음 단계 진행해주세요! 🔥
